### ISSUE TITLE
[discussion] Time Complexity of calculating ITEs in CEVAE

### ISSUE THREAD

**Issue by william-wang-stu** [pyro-ppl/pyro#3331]
### Issue Description

Firstly, I'd like to extend my gratitude for the hard work and dedication you've shown in developing and maintaining Pyro.

I'm new to Pyro. I've noticed that the time complexity of calculating ITEs in CEVAE model is ``O(len(x) * num_samples ** 2)``, which is written in the comments in the code. I've run the inference code of CEVAE on our spark system, and it turns out that the default parameter setting of `num_samples=1000` is too slow. Therefore I've changed it to `num_samples=100`. 

I'm wondering if making such changes will affect the model's inference performace, since the code snippet `with pyro.plate("num_particles", num_samples, dim=-2)` indicates that it will execute `num_samples` times for different sampels. It would be even better if you could explain `pyro.plate` in more details.

Thank you once again for your incredible work on Pyro. I appreciate any guidance or assistance you can provide regarding this issue. I'm looking forward to your response and am more than willing to assist in any further investigation or testing that may be required.

### Environment

'1.8.6+4be5c2e'

### Code Snippet

The code snippet can be found starting from line 601 of file `pyro/contrib/cevae/__init__.py`.
```py
@torch.no_grad()
def ite(self, x, num_samples=None, batch_size=None):
    r"""
    Computes Individual Treatment Effect for a batch of data ``x``.

    .. math::

        ITE(x) = \mathbb E\bigl[ \mathbf y \mid \mathbf X=x, do(\mathbf t=1) \bigr]
               - \mathbb E\bigl[ \mathbf y \mid \mathbf X=x, do(\mathbf t=0) \bigr]

    This has complexity ``O(len(x) * num_samples ** 2)``.

    :param ~torch.Tensor x: A batch of data.
    :param int num_samples: The number of monte carlo samples.
        Defaults to ``self.num_samples`` which defaults to ``100``.
    :param int batch_size: Batch size. Defaults to ``len(x)``.
    :return: A ``len(x)``-sized tensor of estimated effects.
    :rtype: ~torch.Tensor
    """
    if num_samples is None:
        num_samples = self.num_samples
    if not torch._C._get_tracing_state():
        assert x.dim() == 2 and x.size(-1) == self.feature_dim

    dataloader = [x] if batch_size is None else DataLoader(x, batch_size=batch_size)
    logger.info("Evaluating {} minibatches".format(len(dataloader)))
    result = []
    for x in dataloader:
        x = self.whiten(x)
        with pyro.plate("num_particles", num_samples, dim=-2):
            with poutine.trace() as tr, poutine.block(hide=["y", "t"]):
                self.guide(x)
            with poutine.do(data=dict(t=torch.zeros(()))):
                y0 = poutine.replay(self.model.y_mean, tr.trace)(x)
            with poutine.do(data=dict(t=torch.ones(()))):
                y1 = poutine.replay(self.model.y_mean, tr.trace)(x)
        ite = (y1 - y0).mean(0)
        if not torch._C._get_tracing_state():
            logger.debug("batch ate = {:0.6g}".format(ite.mean()))
        result.append(ite)
    return torch.cat(result)
```

**Comment by fritzo** [comment]
You can read about plates [here](*https://pyro.ai/examples/tensor_shapes.html#Declaring-independent-dims-with-plate). The plate should be a vectorized for loop, and hence should be fast.

I don't see where the algorithm is quadratic in `num_samples`. It looks linear in `num_samples` to me. Maybe I'm missing something.

As with any Monte Carlo estimate, variance shrinks as 1/num_samples, so for 100 samples you might see 1 significant digit, for 10000 samples you might see two significant digits.
